desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,882 @@
+    VacuumBoundaryError will be much faster as it avoids the singular virtual casing
+    integral.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    ext_field : MagneticField
+        External field produced by coils.
+    sheet_current : FourierCurrentPotentialField
+        Sheet current on the last closed flux surface. Not required for vacuum fields,
+        but generally needed to correctly solve at finite beta/current. Geometry will
+        be automatically constrained to be the same as the equilibrium LCFS.
+    target : float, ndarray, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        len(target) must be equal to Objective.dim_f
+    bounds : tuple, optional
+        Lower and upper bounds on the objective. Overrides target.
+        len(bounds[0]) and len(bounds[1]) must be equal to Objective.dim_f
+    weight : float, ndarray, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        len(weight) must be equal to Objective.dim_f
+    normalize : bool
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+    s, q : integer
+        Hyperparameters for singular integration scheme, s is roughly equal to the size
+        of the local singular grid with respect to the global grid, q is the order of
+        integration on the local grid
+    src_grid, eval_grid : Grid, optional
+        Collocation grid containing the nodes to evaluate at for source terms and where
+        to evaluate errors.
+    field_grid : Grid, optional
+        Grid used to discretize ext_field.
+    loop : bool
+        If True, evaluate integral using loops, as opposed to vmap. Slower, but uses
+        less memory.
+    name : str
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _linear = False
+    _print_value_fmt = "Boundary Error: {:10.3e} "
+    _units = "(T)"
+    _coordinates = "rtz"
+
+    def __init__(
+        self,
+        eq,
+        ext_field,
+        sheet_current=None,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        s=None,
+        q=None,
+        src_grid=None,
+        eval_grid=None,
+        field_grid=None,
+        loop=True,
+        name="Boundary error",
+    ):
+        if target is None and bounds is None:
+            target = 0
+        self._src_grid = src_grid
+        self._eval_grid = eval_grid
+        self._s = s
+        self._q = q
+        self._ext_field = ext_field
+        self._field_grid = field_grid
+        self._loop = loop
+        self._sheet_current = None
+        things = [eq]
```

Comment:
same note as above on including `ext_field` in things, if it is easy to do now it might be worth doing that in this PR

Reply:
I think that will need some extra testing etc to make sure it works for arbitrary nested collections of `MagneticField` types.

We would also need some non-trivial constraints/objectives for coils/fields etc.